test-data/unit/check-selftype.test

```diff
@@ -0,0 +1,236 @@
+g(A.copy(B()))
+g(B.copy(B()))
+
+[builtins fixtures/bool.pyi]
+
+[case testSelfTypeReturn]
+# flags: --hide-error-context
+from typing import TypeVar, Type
+
+R = TypeVar('R')
+def _type(self: R) -> Type[R]: pass
+
+T = TypeVar('T', bound='A', covariant=True)
+class A:
+    def copy(self: T) -> T:
+        if B():
+            return A()  # E: Incompatible return value type (got "A", expected "T")
+        elif A():
+            return B()  # E: Incompatible return value type (got "B", expected "T")
+        return _type(self)()
+
+class B(A):
+    pass
+
+Q = TypeVar('Q', bound='C', covariant=True)
+class C:
+    def __init__(self, a: int) -> None: pass
+
+    def copy(self: Q) -> Q:
+        if self:
+            return _type(self)(1)
+        else:
+            return _type(self)()  # E: Too few arguments for "C"
+
+
+[builtins fixtures/bool.pyi]
+
+[case testSelfTypeClass]
+# flags: --hide-error-context
+from typing import TypeVar, Type
+
+T = TypeVar('T', bound='A')
+
+class A:
+    @classmethod
+    def new(cls: Type[T]) -> T:
+        return cls()
+
+class B(A):
+    pass
+
+Q = TypeVar('Q', bound='C', covariant=True)
+class C:
+    def __init__(self, a: int) -> None: pass
+
+    @classmethod
+    def new(cls: Type[Q]) -> Q:
+        if cls:
+            return cls(1)
+        else:
+            return cls()  # E: Too few arguments for "C"
+
+
+reveal_type(A.new)  # E: Revealed type is 'def () -> __main__.A'
+reveal_type(B.new)  # E: Revealed type is 'def () -> __main__.B'
+reveal_type(A.new())  # E: Revealed type is '__main__.A'
+reveal_type(B.new())  # E: Revealed type is '__main__.B'
+
+[builtins fixtures/classmethod.pyi]
+
+[case testSelfTypeOverride]
+from typing import TypeVar, cast
+ 
+T = TypeVar('T', bound='A', covariant=True)
+ 
+class A:
+    def copy(self: T) -> T: pass
+
+class B(A):
+    pass
+
+Q = TypeVar('Q', bound='C', covariant=True)
+class C(A):
+    def copy(self: Q) -> Q: pass
+
+reveal_type(C().copy)  # E: Revealed type is 'def () -> __main__.C'
+reveal_type(C().copy())  # E: Revealed type is '__main__.C'
+reveal_type(cast(A, C()).copy)  # E: Revealed type is 'def () -> __main__.A'
+reveal_type(cast(A, C()).copy())  # E: Revealed type is '__main__.A' 
+
+[builtins fixtures/bool.pyi]
+
+[case testSelfTypeSuper]
+# flags: --hide-error-context
+from typing import TypeVar, cast
+ 
+T = TypeVar('T', bound='A', covariant=True)
+ 
+class A:
+    def copy(self: T) -> T: pass
+
+Q = TypeVar('Q', bound='B', covariant=True)
+class B(A):
+    def copy(self: Q) -> Q:
+        reveal_type(self)  # E: Revealed type is 'Q`-1'
+        reveal_type(super().copy)  # E: Revealed type is 'def () -> Q`-1'
+        return super().copy()
+
+[builtins fixtures/bool.pyi]
+
+[case testSelfTypeRecursiveBinding]
+# flags: --hide-error-context
+from typing import TypeVar, Callable, Type
+ 
+T = TypeVar('T', bound='A', covariant=True)
+class A:
+    def copy(self: T, factory: Callable[[T], T]) -> T:
```

Comment:
Using `T` in an argument type in a contravariant position is arguably unsafe, but it also depends on how we understand covariance here. I'll add a separate comment about this in the main PR discussion.